crates/ark/src/variables/variable.rs

```diff
@@ -1864,7 +1864,8 @@ mod tests {
 
     fn inspect_from_expr(code: &str) -> Vec<Variable> {
         let env = Environment::new(harp::parse_eval_base("new.env(parent = emptyenv())").unwrap());
-        env.bind("x".into(), harp::parse_eval_base(code).unwrap());
+        let value = harp::parse_eval_base(code).unwrap();
+        env.bind("x".into(), value.sexp);
```

Comment:
I think what was happening is that with:

```
env.bind("x".into(), harp::parse_eval_base(code).unwrap());
```

- `harp::parse_eval_base(code).unwrap()` creates a `RObject` which does protection
- `env.bind()` consumes `value: impl Into<SEXP>`, and calls `value.into()` on `value`. This consumes the `RObject` (dropping and unprotecting it) and returns the unprotected `SEXP`
- The unprotected `SEXP` is then used in `Rf_defineVar()`

In theory it seems like `Rf_defineVar()` shouldn't do anything to trigger a reallocation, but in practice we see this crash.

---

More broadly I think this suggests a typing issue with `RObject`. I feel like its probably quite bad that we allow the underlying `SEXP` to escape the lifetime of `RObject` so easily. It's like we need to tie the lifetime of `RObject` to the lifetime of underlying `SEXP` unless you call something super explicit like `unsafe into_sexp()` which has a `SAFETY` note that you are taking protection ownership over the `SEXP` from here on out.

Reply:
Removing `impl Into<SEXP>` seems to fix the problem as it seems `temporary()` is only deleted after the function execution. At least that what I can tell from something like:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=dca6608de1df2f3b16ba2c2a217aedf3